[to upstream] Add a LookupRangeCheck trait and minor modifications in preparation for ZSA

ConstanceBeguier opened the pull request:
This PR contains the following modifications:
1. Introduce a `LookupRangeCheck` trait that provides common methods for a lookup range check.
2. Use this new trait as a generic parameter in configs, chips and tests that are using lookup.
3. Add tests for Lookup, ECC, Merkle, and Sinsemilla to verify that the verification key and the proof have not been modified by comparing them to those saved in a file.
4. Add the `EccPointQ` enum to handle public and private points for Q, which is the initialization point for the hashing function.

To adjust existing code to these modifications, developers will just need to import `halo2_gadgets::utilities::lookup_range_check::LookupRangeCheck`.
This [Orchard PR](https://github.com/QED-it/orchard/pull/106) demonstrates how to adapt the Orchard code.

Upstream PR: https://github.com/zcash/halo2/pull/823

ConstanceBeguier committed:
Add variable-base sign-scalar multiplication

ConstanceBeguier committed:
Add hash and blinding_factor functions

ConstanceBeguier committed:
Add some functionalities for MuxChip

ConstanceBeguier committed:
Add commit_from_hash_point

ConstanceBeguier committed:
Add CommitDomain creation from two personalizations

PaulLaux committed:
updated naming

ConstanceBeguier committed:
Add Point::new_from_constant method (#17)

It is now possible to create a Point from a constant.
This functionality is required to evaluate the old nullifier.
- for non split_notes, nf_old = Extract_P([PRF^{nfOrchard}_{nk}(rho_old) + psi_nf) mod q_P] NullifierK + cm_old)
- for split notes, nf_old = Extract_P([PRF^{nfOrchard}_{nk}(rho_old) + psi_nf) mod q_P] NullifierK + cm_old + NullifierL)

ConstanceBeguier committed:
Optimized short range check on 4 and 5 bits (#21)

Short range checks on 4 and 5 bits are now performed with only one lookup (instead of 2).
To do that, we added a column `table_short_range_tag` in the lookup table.
This new column `table_short_range_tag` contains the value
- 4 for rows used in short range check on 4 bits
- 5 for rows used in short range check on 5 bits
- 0 for rows used in short range check on 10 bits

Disable tests on i686 and code coverage in CI

ConstanceBeguier committed:
Add multiplexer chip (#23)

It is now possible to perform a mux between two points or between two non-identity points.
`mux(choice, left, right)` will return `left` when `choice=0` and `right` when `choice=1`.
`choice` must be constrained to `{0, 1}` outside the gate.

It is no longer needed to expose `from_coordinates_unchecked`.

ConstanceBeguier committed:
Add functions to evaluate a Sinsemilla hash from an initial private point (#22)

To share ZEC and ZSA hash computations in Orchard circuit's note commitment evaluation, we need to compute a Sinsemille hash from a private input point.

ConstanceBeguier committed:
Update comments

ConstanceBeguier committed:
Reactivate i686 CI test

ConstanceBeguier committed:
Some minor changes

ConstanceBeguier committed:
Remove q_sinsemilla4_private

ConstanceBeguier committed:
Move mux functionality into CondSwap chip

Commit:
add initial doc

Commit:
clean up code

Commit:
remove cfg[test]

Commit:
remove generator table trait, fix hash_to_point

Commit:
update code for hash

Commit:
fmt

Commit:
fix run tests error

Commit:
minor update

Commit:
test error

Commit:
Revert "test error"

This reverts commit 23a295172b6c8cb0b642512109801b5116d79c6a.

Commit:
re-naming structs and functions

Commit:
fix complex type

Commit:
add round trip test to verify that the pinned verification key (representing the circuit) is as expected

Commit:
update serialized_proof_test_case

Commit:
update

Commit:
update short lookup test

Commit:
cargo fmt

Commit:
fix Clippy error

Commit:
clean up code

Commit:
add initial files

Commit:
minor update

Commit:
remove pub(crate) in ecc

Commit:
add a line after cargo.toml

Commit:
add a line after cargo.toml

Commit:
slide lines in base_field_elem

Commit:
slide lines in base_field_elem

Commit:
adjuct function orders for configure in sinsemilla/chip

Commit:
adjuct function orders for hash_message

Commit:
adjuct function orders for hash_message

Commit:
cleanup code

Commit:
cleanup code

Commit:
cleanup code

Commit:
resolve review

dmidem committed:
Introduce conditionally_save_circuit_to_disk function to reuse the proof saving code in test cases

dmidem committed:
Rename conditionally_save_circuit_to_disk function to conditionally_save_proof_to_disk

dmidem committed:
Introduce test_serialized_proof_with_vk and test_serialized_proof functions to reuse them and so avoid code duplication in serialized_proof_test_case tests

dmidem committed:
Fix cargo clippy error

Commit:
rename

Commit:
fix git error

dmidem committed:
Move all test .bin files and test_circuit.rs module to new tests folder in the root of src

dmidem committed:
Add missed files

dmidem committed:
Move all .bin and vk_ test data files into src/tests folder

Commit:
add tests for short range check

dmidem committed:
Add fixed_verification_key_test_with_circuit function to generalize VK tests like the proof tests are generalized with serialized_proof_test_case_with_circuit

dmidem committed:
Avoid duplicating test data folder name src/tests in every file name, use TEST_DATA_DIR const in test module instead (make the full file name from the fiule names passed, the dir name const and extention (bin for proofs)

dmidem committed:
Remove 'else' in fixed_verification_key_test_with_circuit function, to read after write if env var set and so check the writing

dmidem committed:
Use fs::read, fs::write instad of read_test_case and write_test_case

Commit:
Merge branch 'improve-backward-compatability-without-zsa' of github.com:QED-it/halo2 into improve-backward-compatability-without-zsa

Commit:
revert .github/workflows/ci.yml

Commit:
rename circuit.rs

Commit:
add a line

Commit:
fix ci.yml

dmidem committed:
Rename PallasLookupRC LookupRangeCheck trait alias to PallasLookup. Add and use PallasLookupConfig type alias for LookupRangeCheckConfigspecialized with pallas::Base and sinsemilla::K

dmidem committed:
Move circuit test data to tests/circuit_reference_data folder, fix names of the test data files to follow the same convention: .proof.bin for proof tests, .vk.txt for verifying key/layout tests

dmidem committed:
Rename circuit_reference_data to circuit_data

dmidem committed:
Fix cargo fmt issues

dmidem committed:
Try to increase cargo-tarpaulin timeout in ci.yml from 600 to 1800 to see if it helps to resolve test failing on GitHub

dmidem committed:
Add default generic arguments for PallasLookup to improve backward compatibility

This change adds default generic arguments for types implementing the PallasLookup trait,
assigning them to PallasLookupConfig by default. This adjustment allows existing code to
function without requiring users to explicitly specify PallasLookupConfig, thus minimizing
breaking changes. Users will still need to bring the PallasLookup trait into scope using
'use ... PallasLookup'.

dmidem committed:
Try to increase cargo-tarpaulin timeout in ci.yml from to 3600

Commit:
update

Commit:
cargo fmt

dmidem committed:
Merge branch 'improve-backward-compatability-without-zsa-default-generic-arg' into improve-backward-compatability-without-zsa

Commit:
add index to proofs

Commit:
rename PallasLookupRC10b to PallasLookupRCConfig

Commit:
resolve some comments

Commit:
Fix error in ci.yml

Commit:
Merge branch 'improve-backward-compatability-without-zsa' of github.com:QED-it/halo2 into improve-backward-compatability-without-zsa

ConstanceBeguier committed:
Cleanup code and some renamings

ConstanceBeguier committed:
rust fmt

ConstanceBeguier committed:
Put together test_against_stored_vk/proof into test_against_stored_circuit

Previously, the functions test_against_stored_vk and test_against_stored_proof
each generated a verification key (vk). This operation can be quite slow.
We are combining these two tests to generate the vk only once.

ConstanceBeguier committed:
Refactor

ConstanceBeguier committed:
Remove MerkleSinsemillaInstructions

ConstanceBeguier committed:
Remove small functions in SinsemillaChip

ConstanceBeguier committed:
Refactor in lookup_range_check

ConstanceBeguier committed:
Rename PallasLookupRC(Config)

PallasLookupRangeCheck <- PallasLookupRC
PallasLookupRangeCheckConfig <- PallasLookupRCConfig

ConstanceBeguier committed:
Add Lookup generic argument in ecc tests

ConstanceBeguier committed:
Add Lookup generic argument in Sinsemilla tests

ConstanceBeguier committed:
Refactor

ConstanceBeguier committed:
Add Lookup generic argument in Lookup tests

ConstanceBeguier committed:
Add Lookup generic argument into mul_fixed/short tests

ConstanceBeguier committed:
Add Lookup generic argument into Merkle tests

ConstanceBeguier committed:
Refactor

ConstanceBeguier committed:
Add generic Lookup argument in short_range_check tests

ConstanceBeguier committed:
Update EccPointQ enum

ConstanceBeguier committed:
Update EccPointQ enum

PaulLaux commented on halo2_gadgets/src/ecc.rs:
```diff
@@ -895,12 +908,24 @@ pub(crate) mod tests {
     fn ecc_chip() {
-        let k = 13;
-        let circuit = MyCircuit { test_errors: true };
+        let k = 11;
```
Why this change? maybe there is a reason for k=13

PaulLaux commented on halo2_gadgets/src/ecc.rs:
```diff
@@ -895,12 +908,24 @@ pub(crate) mod tests {
+        let k = 11;
+        let circuit: MyCircuit<PallasLookupRangeCheckConfig> = MyCircuit {
+            test_errors: true,
+            _lookup_marker: PhantomData,
```
is there a way to avoid this ugly thing on init? maybe an init function

PaulLaux commented on halo2_gadgets/src/sinsemilla.rs:
```diff
@@ -516,19 +515,25 @@ pub(crate) mod tests {
         fn without_witnesses(&self) -> Self {
-            MyCircuit {}
+            MyCircuit {
+                _lookup_marker: PhantomData,
```
same

PaulLaux commented on halo2_gadgets/src/sinsemilla.rs:
```diff
@@ -516,19 +515,25 @@ pub(crate) mod tests {
     }
 
-    struct MyCircuit {}
+    struct MyCircuit<Lookup: PallasLookupRangeCheck> {
+        _lookup_marker: PhantomData<Lookup>,
+    }
```
consider adding
```rust
impl<Lookup: PallasLookupRangeCheck> MyCircuit<Lookup> {                                                                                                                                                                                      
     fn new() -> Self {                                                                                                                                                                                                                        
         MyCircuit {                                                                                                                                                                                                                           
             _lookup_marker: PhantomData,                                                                                                                                                                                                                                                                                                                                                                                        
         }                                                                                                                                                                                                                                     
     }                                                                                                                                                                                                                                         
 }        
 ```
or something similar

PaulLaux commented on halo2_gadgets/src/sinsemilla/chip/hash_to_point.rs:
```diff
@@ -374,6 +386,80 @@ where
 
         Ok((x_a, y_a, zs))
     }
+
```
can we move it back so it does not look like new code in the diff?

PaulLaux commented on halo2_gadgets/src/utilities/lookup_range_check.rs:
```diff
@@ -397,254 +468,304 @@ mod tests {
-            type Config = LookupRangeCheckConfig<F, K>;
-            type FloorPlanner = SimpleFloorPlanner;
+    #[derive(Clone, Copy)]
+    struct MyLookupCircuit<F: PrimeFieldBits, Lookup: LookupRangeCheck<F, K>> {
```
same for init function

PaulLaux commented on halo2_gadgets/src/utilities/lookup_range_check.rs:
```diff
@@ -397,254 +468,304 @@ mod tests {
+    struct MyLookupCircuit<F: PrimeFieldBits, Lookup: LookupRangeCheck<F, K>> {
+        num_words: usize,
+        _field_marker: PhantomData<F>,
+        _lookup_marker: PhantomData<Lookup>,
```
Can we have them both in one? `_marker: PhantomData<<F,Lookup>>` or similar?

ConstanceBeguier replied:
_marker: PhantomData<(F, Lookup)>

PaulLaux reviewed (commented):
Added some minor comments and questions

ConstanceBeguier committed:
Take into account Pablo's review

ConstanceBeguier committed:
Remove unnecessary changes in CI

ConstanceBeguier committed:
Merge branch 'main' into improve-backward-compatability-without-zsa

ConstanceBeguier committed:
Take into account code review

ConstanceBeguier committed:
Fix CI

- Fix bitrot CI
- Fix clippy beta
- Rename our tests module by test_circuits (and the corresponding files and folders)

ConstanceBeguier committed:
Take into account code review

ConstanceBeguier committed:
Simplify check_hash_result function

ConstanceBeguier committed:
Simplify EccPointQ enum
